### PR TITLE
Minimal resolver

### PR DESCRIPTION
**Pull request by dim**
@mxmCherry check out this (instead of #3)

### CODE REVIEW

File: bfs.go [hunk 1]
```diff
@@ -56,3 +59,36 @@ type Iterator interface {
 type supportsCopying interface {
 	Copy(context.Context, string, string) error
 }
+
+// --------------------------------------------------------------------
+
+// Resolve opens a bucket from a URL string.
```
Comment (mxmCherry) [inline bfs.go]: Minor: doc contradicts with impl - it's not a "url string", but already parsed URL.

File: bfs.go [hunk 2]
```diff
@@ -56,3 +59,36 @@ type Iterator interface {
...
+func Resolve(ctx context.Context, u *url.URL) (Bucket, error) {
+	registryLock.Lock()
```
Comment (mxmCherry) [inline bfs.go]: Minor: that's fine, but I'd not overcomplicate and just tell in README (or better in godoc comment) that's it's not thread safe and should be called only in `func init()`. Or you expect these to be concurrent as well?

Comment (mxmCherry) [reply bfs.go]: Also there's a sync.Map, as I remember - probs, worth being used here to get rid of mutex.

File: bfs.go (hunk 1 above)
Comment (mxmCherry) [reply bfs.go]: Also, the goal is to save some repetitive code, so accepting a string would be more convenient.

File: bfs.go [hunk 3]
```diff
@@ -56,3 +59,36 @@ type Iterator interface {
...
+// Resolver constructs a bucket from a URL.
+type Resolver func(context.Context, *url.URL) (Bucket, error)
+
+var (
+	registry     = map[string]Resolver{}
+	registryLock sync.Mutex
+)
+
+// RegisterProtocol registers a new protocol with a resolver.
+func RegisterProtocol(scheme string, resv Resolver) {
```
Comment (mxmCherry) [inline bfs.go]: Ultra-minor - I'd call it just `Register` (as I did), or rename `schema` to `proto` as well... But really, really minor.

File: bfsfs/bfsfs.go [hunk 4]
```diff
@@ -2,13 +2,22 @@
...
 	"os"
 	"path/filepath"
 
 	"github.com/bsm/bfs"
 )
 
+func init() {
+	bfs.RegisterProtocol("file", func(_ context.Context, u *url.URL) (bfs.Bucket, error) {
+		q := u.Query()
+		return New(u.Path, q.Get("tmpdir"))
```
Comment (mxmCherry) [inline bfsfs/bfsfs.go]: Oh, that's really neat!

File: bfsgs/bfsgs.go [hunk 5]
```diff
@@ -12,17 +13,25 @@ import (
...
 )
 
-type gsBucket struct {
-	bucket *storage.BucketHandle
-	config *Config
+func init() {
+	bfs.RegisterProtocol("gs", func(ctx context.Context, u *url.URL) (bfs.Bucket, error) {
+		q := u.Query()
+		return New(ctx, u.Host, &Config{
+			Prefix: q.Get("prefix"),
```
Comment (mxmCherry) [inline bfsgs/bfsgs.go]: Config is tiny right now, but what if it grows (gains new fields)? Still looks fine to me: don't like default impl - register yours. But I'd remove `panic` on double-register then (very minor).

File: bfs.go (hunk 1 above)
Comment (dim) [reply bfs.go]: @mxmCherry I thought about passing a string first, but that's not really helpful if you want to use the URL. `Resolve` would have to parse the string and return the `URL` which is an inconvenience and not exactly what I was after with "minimal"

File: bfs.go (hunk 2 above)
Comment (dim) [reply bfs.go]: I don't see a reason to ditch the mutex, it's simple and easy. also package loading in Go is not deterministic, if someone does:

```
import (
 _ "github.com/bfs/bfss3"
 _ "github.com/bfs/bfsgs"
)
```

Can anyone be absolutely sure that they are not loaded in parallel? Better: keep mutex!

File: bfs.go (hunk 3 above)
Comment (dim) [reply bfs.go]: was following https://golang.org/pkg/net/http/#Transport.RegisterProtocol

**Comment by mxmCherry** [comment]
I missed that path is not returning...

So the usage for this is like:

```
u, err := url.Parse("file:///home/user/Documents/my.doc")
if err != nil {
  panic(err)
}

bkt, err := bfs.Resolve(context.TODO(), u)
if err != nil {
  panic(err)
}

f, err := bkt.Open(context.TODO(), u.Path)
...
```

?

IMO, accepting string url and returning `bucket, path, err` would be more convenient, but Ok.

**Pull request merged**